### PR TITLE
PoC of Voucherify Integration 

### PR DESCRIPTION
**Pull request by p-zielinski**
Please see README-VOUCHERIFY.md for more details

### CODE REVIEW

**Review by michalbujalski — APPROVED** [review]

File: src/components/Cart/cart.js [hunk 1]
```diff
@@ -51,8 +52,25 @@ export const PriceExcludeVAT1 = ({ price, caption }) => {
   )
 }
 
-export const CartMobileItem = ({ cartItem }) => {
-  const { incrementCartItemQty, decrementCartItemQty } = useCart()
+export const CartMobileItem = ({ cartItem, cart }) => {
+  const { changeCartItemQty } = useCart()
+  const discountsDetails = cart?.metadata?.mixins?.discountsDetails || []
```
Comment (lstypka) [inline src/components/Cart/cart.js:57]: Please change the line. to
```
  const discountsDetails = cart?.mixins?.voucherify?.discountsDetails || []
````

File: src/context/cart-provider.js [hunk 2]
```diff
@@ -104,6 +104,10 @@ const CartProvider = ({ children }) => {
     return cartAccount.discounts
   }, [cartAccount.discounts])
 
+  const mixins = useMemo(() => {
+    return cartAccount.metadata?.mixins || {}
+  }, [cartAccount.metadata?.mixins])
+
```
Comment (lstypka) [inline src/context/cart-provider.js:110]: please change the function to
```
  const mixins = useMemo(() => {
    return cartAccount?.mixins || {}
  }, [cartAccount?.mixins])
```

File: src/integration/buildIntegrationCartFromEmporixCart.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    }) || []
+  const newCodesObjects =
+    newCodes?.map((code) => {
+      return {
+        code,
+        status: 'NEW',
+      }
+    }) || []
+  const currentlyAppliedCoupons = Array.isArray(
```
Comment (lstypka) [inline src/integration/buildIntegrationCartFromEmporixCart.js:62]: Please change the following four lines to:
```
    emporixCart?.mixins?.voucherify?.appliedCoupons
  )
    ? emporixCart?.mixins?.voucherify?.appliedCoupons
    : []
```

File: src/integration/buildIntegrationCartFromEmporixCart.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    'code'
+  )
+
+  return {
+    id: emporixCart.id,
+    customer: mapEmporixUserToVoucherifyCustomer(
+      customer,
+      customerAdditionalMetadata
+    ),
+    sessionKey: emporixCart.metadata?.mixins?.sessionKey,
```
Comment (lstypka) [inline src/integration/buildIntegrationCartFromEmporixCart.js:96]: please change the line to 
```
    sessionKey: emporixCart?.mixins?.voucherify?.sessionKey,
```

File: src/integration/emporix/emporixApi.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        maxRedemptions: 1,
+        categoryRestricted: false,
+        issuedTo: emporixCart.customerId,
+      }),
+    }
+  )
+  return await couponRaw.json()
+}
+
+export const updateCartMetadataMixins = async (
```
Comment (lstypka) [inline src/integration/emporix/emporixApi.js:116]: Please change the function according to this snippet
```
export const updateCartMetadataMixins = async (
  emporixCart,
  validationResult,
  deletedCodes,
  discountsDetails
) => {
  const { newSessionKey, applicableCoupons, availablePromotions } =
    validationResult
  const mixinsSchema = {
    ...(emporixCart?.metadata?.mixins || {})
  }
  mixinsSchema['voucherify'] = "https://res.cloudinary.com/saas-ag/raw/upload/v1686167929/voucherify/voucherify.json"
  
  const mixins = {
    ...(emporixCart?.mixins || {}),
    voucherify : {
      sessionKey: newSessionKey || emporixCart.mixins?.voucherify?.sessionKey,
      appliedCoupons: applicableCoupons
        .filter((coupon) => {
          if (!deletedCodes) {
            return true
          }
          return !deletedCodes.includes(coupon.id)
        })
        .map((coupon) => {
          return {
            code: coupon.id,
            status: 'APPLIED',
            type: coupon.object,
            banner: coupon.banner,
          }
        }),
      availablePromotions,
      discountsDetails,
    }
  }
  const emporixAccessToken = await getEmporixAPIAccessToken()
  const cartUpdate = await fetch(
    `${process.env.REACT_APP_API_URL}/cart/${localStorage.getItem(
      TENANT
    )}/carts/${emporixCart.id}`,
    {
      method: 'PUT',
      headers: {
        Authorization: `Bearer ${emporixAccessToken}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
          mixins,
          metadata : {
            mixins : mixinsSchema
          }
      }),
    }
  )
  if (cartUpdate.status !== 204) {
    throw {
      error: `Could not update cart with id: ${emporixCart.id}`,
    }
  }
}
```

File: src/pages/cart/CartTable.js [hunk 6]
```diff
@@ -31,108 +34,219 @@ const CartProductInfo = ({ cart }) => {
...
+    cartList.map((cartItem, index) => {
+      const itemId = cartItem.itemYrn?.split?.(';')?.at?.(-1)
+      if (itemId === open) {
+        newName = cartItem?.product?.name
+      }
+    })
+    setProductName(newName)
+  }, [open])
+
+  const discountsDetails = cart?.metadata?.mixins?.discountsDetails || []
```
Comment (lstypka) [inline src/pages/cart/CartTable.js:60]: please change the line to:
```
 const discountsDetails = cart?.mixins?.voucherify?.discountsDetails || []
```

File: src/pages/shared/Qualification.js [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+      (qualification) => qualification?.code !== code
+    )
+    setUsersSavedQualificationsState(updatedUsersSavedQualifications)
+    setUsersSavedQualifications(user, updatedUsersSavedQualifications)
+  }
+
+  const [isBeingApplied, setIsBeingApplied] = useState(false)
+  const [areProductsBeingAdded, setAreProductsBeingAdded] = useState(false)
+  const { applyPromotion, applyDiscount, cartAccount, recheckCart } = useCart()
+  const cartMixins = cartAccount.metadata?.mixins || {}
```
Comment (lstypka) [inline src/pages/shared/Qualification.js:86]: please change it to:
```
const cartMixins = cartAccount?.mixins?.voucherify || {}
 ```

**Review by lstypka — CHANGES_REQUESTED** [review]

File: src/integration/emporix/emporixApi.js [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+      TENANT
+    )}/carts/${emporixCart.id}`,
+    {
+      method: 'PUT',
+      headers: {
+        Authorization: `Bearer ${emporixAccessToken}`,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        metadata: {
```
[the remark this answers is not in the document]
Comment (p-zielinski) [reply src/integration/emporix/emporixApi.js:166]: Yes, I had to missed that :/
Thank You so much!

**Review by lstypka — APPROVED** [review]

**Pull request merged**